### PR TITLE
Azure timeout error

### PR DESCRIPTION
**Pull request by garuppel**
Azure times out after about 60 minutes.

### CODE REVIEW

**Comment by garuppel** [comment]
At the moment we can't test the discovery.Start() method any further, we have to wait for the Issue #473  to be resolved first.

File: service/discovery/discovery_test.go [hunk 1]
```diff
@@ -172,7 +202,193 @@ func TestQuery(t *testing.T) {
...
+
+			for _, env := range tt.fields.envVariables {
+				if env.hasEnvVariable {
+					t.Setenv(env.envVariableKey, env.envVariableValue)
+				}
+			}
+
+			// We need to logout from Azure to check if no authorizer is present
+			if tt.fields.azLogout {
+				cmd := exec.Command("az", "logout")
```
Comment (oxisto) [inline service/discovery/discovery_test.go:317]: Hmm I am not sure if we really want to do this. I am not a big fan of unit tests that change the behaviour of the system we are running on.

File: service/discovery/discovery_test.go [hunk 2]
```diff
@@ -172,7 +202,193 @@ func TestQuery(t *testing.T) {
...
+	}{
+		{
+			name: "Azure authorizer from file",
+			fields: fields{
+				hasRPCConnection: true,
+				envVariables: []envVariable{
+					{
+						hasEnvVariable:   true,
+						envVariableKey:   "AZURE_AUTH_LOCATION",
+						envVariableValue: "service/discovery/testfile/credentials_test_file",
```
Comment (oxisto) [inline service/discovery/discovery_test.go:236]: Do we already have a folder for test files somewhere? Or maybe rename the directory `testfile` to testing?

File: service/discovery/discovery_test.go (hunk 1 above)
Comment (garuppel) [reply service/discovery/discovery_test.go:317]: Ah, sorry, that was a first try. I've already do the test without the az logout and forgot to delete it.

File: service/discovery/discovery.go [hunk 3]
```diff
@@ -90,38 +91,63 @@
...
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "could not initialize stream to Assessment: %v", err)
+		}
 	}
 
-	// create an authorizer from env vars or Azure Managed Service Identity
-	authorizer, err := auth.NewAuthorizerFromCLI()
+	// create an authorizer from file or as fallback from the CLI
+	// if authorizer is from CLI, the access token expires after 75 minutes
+	authorizer, err := auth.NewAuthorizerFromFile(autorest_azure.PublicCloud.ResourceManagerEndpoint)
```
Comment (lebogg) [inline service/discovery/discovery.go:133]: Maybe extract following authentication code in a separate method and like I wrote above, maybe put it in ./azure/azure.go?

File: service/discovery/discovery.go [hunk 4]
```diff
@@ -90,38 +91,63 @@
...
+		log.Info("Using Azure authorizer from CLI. The discovery times out after 1 hour.")
+	} else {
+		log.Info("Using Azure authorizer from file.")
 	}
 
 	k8sClient, err := k8s.AuthFromKubeConfig()
 	if err != nil {
 		log.Errorf("Could not authenticate to Kubernetes: %v", err)
-		return nil, err
+		return nil, status.Errorf(codes.FailedPrecondition, "could not authenticate to Kubernetes: %v", err)
```
Comment (lebogg) [inline service/discovery/discovery.go:150]: Nice, I like using that status code

File: service/discovery/aws/compute.go [hunk 5]
```diff
@@ -68,8 +68,8 @@ var newFromConfigEC2 = ec2.NewFromConfig
 // newFromConfigLambda holds lambda.NewFromConfig(...) allowing a test function tp mock it
 var newFromConfigLambda = lambda.NewFromConfig
 
-// NewComputeDiscovery constructs a new awsS3Discovery initializing the s3-virtualMachineAPI and isDiscovering with true
-func NewComputeDiscovery(client *Client) discovery.Discoverer {
+// NewAwsComputeDiscovery constructs a new awsS3Discovery initializing the s3-virtualMachineAPI and isDiscovering with true
+func NewAwsComputeDiscovery(client *Client) discovery.Discoverer {
```
[the remark this answers is not in the document]
Comment (garuppel) [reply service/discovery/aws/compute.go:72]: Yes, but in the discovery.Start() I found that the following and wanted to unify: 
```
aws.NewAwsStorageDiscovery(awsClient),
aws.NewComputeDiscovery(awsClient),
```

File: service/discovery/azure/azure.go [hunk 6]
```diff
@@ -83,17 +83,29 @@
 		return errors.New("no authorized was available")
 	}
 
-	// for now, do not re-authorize. in the future, we would probably need to check, if
-	// the token is still valid. or maybe Azure does this for us?
+	// If using NewAuthorizerFromFile() in discovery file, we do not need to re-authorize.
+	// If using NewAuthorizerFromCLI() in discovery file, the token expires after 75 minutes.
+	// TODO: How do we check, if the token is still valid?
```
[the remark this answers is not in the document]
Comment (garuppel) [reply service/discovery/azure/azure.go:88]: Yes... no... I don't know how to proceed. The problem is that we have to do the token refresh ourselves if we use the auth.NewAuthorizerFromCLI() (currently it is only a fallback).

File: service/discovery/aws/compute_test.go [hunk 7]
```diff
@@ -437,8 +437,8 @@ func TestComputeDiscovery_NewComputeDiscovery(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if got := NewComputeDiscovery(tt.args.client); !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("NewComputeDiscovery() = %v, want %v", got, tt.want)
+			if got := NewAwsComputeDiscovery(tt.args.client); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("NewAwsComputeDiscovery() = %v, want %v", got, tt.want)
```
[the remark this answers is not in the document]
Comment (garuppel) [reply service/discovery/aws/compute_test.go:441]: I would leave it as it is.

File: service/discovery/testfile/credentials_test_file [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
```
[the remark this answers is not in the document]
Comment (garuppel) [reply service/discovery/testfile/credentials_test_file]: We could also name it testdata:  [link](https://github.com/Azure/go-autorest/tree/master/autorest/azure/auth)

File: service/discovery/azure/azure.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (lebogg) [reply service/discovery/azure/azure.go:93]: Then I would just remove the TODO. Maybe  leave it as a comment but it is stated in the line above anyway.

File: service/discovery/aws/compute.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (lebogg) [reply service/discovery/aws/compute.go:72]: Good point. Yea, then we leave it like that for now. We can address it in the consistent code style issue

File: service/discovery/testfile/credentials_test_file (hunk 8 above)
[the remark this answers is not in the document]
Comment (garuppel) [reply service/discovery/testfile/credentials_test_file]: At the moment we only have one testfile. And I would prefer to create multiple subfolders if needed. So, for now, I let it like that. :-)

File: service/discovery/azure/azure.go [hunk 9]
```diff
@@ -83,17 +83,29 @@
...
 		return
 	}
 
 	subClient := subscriptions.NewClient()
 	a.apply(&subClient.Client)
 
+	// get subscriptions
+	page, err := subClient.List(context.Background())
+	if err != nil {
+		return
```
[the remark this answers is not in the document]
Comment (garuppel) [reply service/discovery/azure/azure.go:105]: I don't know where to get the tenant id at that point.

[the remark this answers is not in the document]
Comment (lebogg) [reply service/discovery/azure/azure.go:104]: No, I didn't mean to add the tenant ID. Just to add an error message, maybe w/o tenant at all. Just for indicating which function went wrong

[the remark this answers is not in the document]
Comment (lebogg) [reply service/discovery/azure/azure.go:104]:
```suggestion
		return fmt.Errorf("could not get subscriptions: %v", err)
```

**Pull request merged**